platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -198,62 +145,142 @@ public Hash hash(
+                } else {
+                    int len = 1 << height;
+                    long rankPath = path;
+                    for (int i = 0; i < height; i++) {
+                        rankPath = Path.getLeftChildPath(rankPath);
+                    }
+                    while (len > 1) {
+                        for (int i = 0; i < len / 2; i++) {
+                            final long hashedPath = Path.getParentPath(rankPath + i * 2);
+                            Hash left = ins[i * 2];
+                            Hash right = ins[i * 2 + 1];
+                            if ((left == null) && (right == null)) {
+                                ins[i] = null;
+                            } else {
+                                if (left == null) {
+                                    left = hashReader.apply(rankPath + i * 2);
+                                }
+                                if (right == null) {
+                                    right = hashReader.apply(rankPath + i * 2 + 1);
+                                }
+                                ins[i] = hash(hashedPath, left, right);
+                                listener.onNodeHashed(hashedPath, ins[i]);
+                            }
+                        }
+                        rankPath = Path.getParentPath(rankPath);
+                        len = len >> 1;
+                    }
+                    hash = ins[0];
+                }
+                out.setHash(getIndexInOut(), hash);
+                return true;
+            } catch (final Throwable e) {
+                completeExceptionally(e);
+                throw e;
+            }
+        }
+
+        static Hash hash(final long path, final Hash left, final Hash right) {
+            final long classId = path == ROOT_PATH ? VirtualRootNode.CLASS_ID : VirtualInternalNode.CLASS_ID;
+            final int serId = path == ROOT_PATH
+                    ? VirtualRootNode.ClassVersion.CURRENT_VERSION
+                    : VirtualInternalNode.SERIALIZATION_VERSION;
+            final HashBuilder builder = HASH_BUILDER_THREAD_LOCAL.get();
+            builder.reset();
+            builder.update(classId);
+            builder.update(serId);
+            builder.update(left);
+            builder.update(right);
+            return builder.build();
+        }
+
+        private int getIndexInOut() {
+            if (path == 0) {
+                return 0;
+            }
+            final long firstInPathInOut = Path.getLeftGrandChildPath(out.path, out.height);
+            return (int) (path - firstInPathInOut);
+        }
+    }
+
     public Hash hash(
             final LongFunction<Hash> hashReader,
             final Iterator<VirtualLeafRecord<K, V>> sortedDirtyLeaves,
             final long firstLeafPath,
             final long lastLeafPath,
             VirtualHashListener<K, V> listener) {
 
-        // Hashing algo v5. This version of the hashing algorithm is designed to optimally process
```

Comment:
It would be nice to have some text describing how this new implementation actually works.

Reply:
Some comments are provided